### ISSUE TITLE
Network security group null value is still creating the resource.

### ISSUE THREAD

**Issue by anoopkrrish07** [Azure/terraform-azurerm-compute#261]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Greenfield/Brownfield provisioning

greenfield

### Terraform Version

1.4.6

### Module Version

5.2.2

### AzureRM Provider Version

3.55.0

### Affected Resource(s)/Data Source(s)

azure vm

### Terraform Configuration Files

```hcl
main.tf
=======

data "azurerm_resource_group" "vnet-rg" {
  name = var.vnet_resourcegroup_name
}

data "azurerm_virtual_network" "vnet" {
  name                = var.vnet_name
  resource_group_name = data.azurerm_resource_group.vnet-rg.name
}

module "windowsservers" {
  source              = "Azure/compute/azurerm"
  version             = "5.2.2"
  vm_hostname         = "${random_id.rg_name.hex}-w10" # line can be removed if only one VM module per resource group
  resource_group_name = azurerm_resource_group.rg_vmwin.name
  vnet_subnet_id      = var.vnet_subnet_id
  location            = var.azure_location
  vm_size             = var.vm_size
  is_windows_image    = true
  admin_username      = var.admin_username
  admin_password      = var.admin_password
  allocation_method   = var.allocation_method
  availability_set_enabled         = false
  enable_accelerated_networking    = true
  network_security_group           = var.network_security_group
  delete_os_disk_on_termination = var.delete_os_disk_on_termination
  delete_data_disks_on_termination = var.delete_data_disks_on_termination
  nb_public_ip                     = var.nb_public_ip
  vm_os_publisher                  = var.vm_os_publisher
  vm_os_offer                      = var.vm_os_offer
  vm_os_sku                        = var.vm_os_sku
  vm_os_version                    = var.vm_os_version
  storage_account_type             = var.storage_account_type


  depends_on = [ azurerm_resource_group.rg_vmwin ]
}

variable.tf
===========

#-----------------------------------------------------------------------------------------------------------
# Global variables
#-----------------------------------------------------------------------------------------------------------

variable "azure_location" {
  type    = string
  default = "uksouth"
}

#-----------------------------------------------------------------------------------------------------------
# Naming Convention variables
#-----------------------------------------------------------------------------------------------------------

variable "resource_type" {
  type    = string
  default = "vm"  
}

variable "env" {
  type    = string
  default = "poc" 
}

variable "app" {
  type    = string
  default = "app" 
}

variable "location_prefix" {
  type    = string
  default = "seu"
}


#-----------------------------------------------------------------------------------------------------------
# azure creds
#-----------------------------------------------------------------------------------------------------------

variable "subscription_id" {
  type    = string
  description = "Stackgenie Azure subscription id" 
}

variable "client_id" { 
  type    = string
  description = "Stackgenie Azure" 
}

variable "client_secret" {
  type    = string
  description = "Stackgenie Azure"
}

variable "tenant_id" {
  type    = string
  description = "Stackgenie Azure" 
}

#-----------------------------------------------------------------------------------------------------------
# VM Variables
#-----------------------------------------------------------------------------------------------------------

variable "vnet_resourcegroup_name" {
  type = string
  description = "vnet resource group name"
}

variable "vnet_name" {
  type = string
  description = "vnet name"
}

variable "vnet_subnet_id" {
  type        = string
  description = "The subnet id of the virtual network where the virtual machines will reside."
}

variable "admin_password" {
  type        = string
  default     = ""
  description = "The admin password to be used on the VMSS that will be deployed. The password must meet the complexity requirements of Azure."
}

variable "admin_username" {
  type        = string
  default     = "azureuser"
  description = "The admin username of the VM that will be deployed."
}

variable "allocation_method" {
  type        = string
  default     = "Dynamic"
  description = "Defines how an IP address is assigned. Options are Static or Dynamic."
}

variable "delete_data_disks_on_termination" {
  type        = bool
  default     = false
  description = "Delete data disks when machine is terminated."
}

variable "delete_os_disk_on_termination" {
  type        = bool
  default     = false
  description = "Delete OS disk when machine is terminated."
}

variable "enable_accelerated_networking" {
  type        = bool
  default     = false
  description = "(Optional) Enable accelerated networking on Network interface."
}

variable "identity_ids" {
  type        = list(string)
  default     = []
  description = "Specifies a list of user managed identity ids to be assigned to the VM."
}

variable "identity_type" {
  type        = string
  default     = ""
  description = "The Managed Service Identity Type of this Virtual Machine."
}

variable "license_type" {
  type        = string
  default     = null
  description = "Specifies the BYOL Type for this Virtual Machine. This is only applicable to Windows Virtual Machines. Possible values are Windows_Client and Windows_Server"
}

variable "nb_public_ip" {
  type        = number
  default     = 0
  description = "Number of public IPs to assign corresponding to one IP per vm. Set to 0 to not assign any public IP addresses."
}

variable "storage_account_type" {
  type        = string
  default     = "Premium_LRS"
  description = "Defines the type of storage account to be created. Valid options are Standard_LRS, Standard_ZRS, Standard_GRS, Standard_RAGRS, Premium_LRS."
}

variable "storage_os_disk_size_gb" {
  type        = number
  default     = null
  description = "(Optional) Specifies the size of the data disk in gigabytes."
}

variable "vm_hostname" {
  type        = string
  default     = "myvm"
  description = "local name of the Virtual Machine."
}

# variable "vm_os_id" {
#   type        = string
#   default     = ""
#   description = "The resource ID of the image that you want to deploy if you are using a custom image.Note, need to provide is_windows_image = true for windows custom images."
# }

variable "vm_os_offer" {
  type        = string
  default     = ""
  description = "The name of the offer of the image that you want to deploy. This is ignored when vm_os_id or vm_os_simple are provided."
}

variable "vm_os_publisher" {
  type        = string
  default     = ""
  description = "The name of the publisher of the image that you want to deploy. This is ignored when vm_os_id or vm_os_simple are provided."
}

variable "vm_os_sku" {
  type        = string
  default     = ""
  description = "The sku of the image that you want to deploy. This is ignored when vm_os_id or vm_os_simple are provided."
}

variable "vm_os_version" {
  type        = string
  default     = "latest"
  description = "The version of the image that you want to deploy. This is ignored when vm_os_id or vm_os_simple are provided."
}

variable "vm_size" {
  type        = string
  default     = "Standard_D2s_v3"
  description = "Specifies the size of the virtual machine."
}

# variable "ssh_key" {
#   type        = string
#   default     = "~/.ssh/id_rsa.pub"
#   description = "Path to the public key to be used for ssh access to the VM. Only used with non-Windows vms and can be left as-is even if using Windows vms. If specifying a path to a certification on a Windows machine to provision a linux vm use the / in the path versus backslash.e.g. c : /home/id_rsa.pub."
# }

# variable "enable_ssh_key" {
#   type        = bool
#   default     = false
#   description = "(Optional) Enable ssh key authentication in Linux virtual Machine."
# }

variable "is_windows_image" {
  type        = bool
  default     = false
  description = "Boolean flag to notify when the custom image is windows based."
}

variable "availability_set_enabled" {
  type        = bool
  default     = true
  description = "(Optional) Enable or Disable availability set.  Default is `true` (enabled)."
  nullable    = false
}

variable "network_security_group" {
  type = object({
    id = string
  })
  default     = null
  description = "The network security group we'd like to bind with virtual machine. Set this variable will disable the creation of `azurerm_network_security_group` and `azurerm_network_security_rule` resources."

  validation {
    condition     = var.network_security_group == null ? true : var.network_security_group.id != null
    error_message = "When `var.network_security_group` is not `null`, `var.network_security_group.id` is required."
  }
}

## resource-group.tf
====================

resource "random_id" "rg_name" {
  byte_length = 4
}

resource "azurerm_resource_group" "rg_vmwin" {
  location = var.azure_location
  name     = "vmwin10-${var.env}-rg-${var.location_prefix}-${random_id.rg_name.hex}"

}

## provider.tf
==============

terraform {
  required_providers {
    azurerm = {
      source = "hashicorp/azurerm"
      version = "~> 3.55.0"
    }
  }
  backend "azurerm" {
  }  
}

provider "azurerm" {
  # Configuration options
  features {}
}
```

### tfvars variables values

```hcl
## VNET Parameters
vnet_resourcegroup_name          = "vnet-rg-name"
vnet_name                        = "vnet-name"
#private_subnet                  = "private-subnet-1"
vnet_subnet_id                   = "private-subnet-id"
admin_username                   = "azuser"
admin_password                   = "password"
vm_size                          = "Standard_D2s_v3" 
delete_data_disks_on_termination = true
delete_os_disk_on_termination    = true
vm_os_publisher                  = "MicrosoftWindowsDesktop"
vm_os_offer                      = "Windows-10"
vm_os_sku                        = "win10-22h2-pron-g2"
vm_os_version                    = "latest"
network_security_group           = null
```

### Debug Output/Panic Output

```shell
https://gist.github.com/anoopkrrish07/a230dc0cdc3666fadc9622e119de264e
```

### Expected Behaviour

By setting the "**network_security_group**" to "**null**", that should not create a new network security group. I was adhering to Azure's best practices by utilizing an existing network security group that's already associated at the subnet level.

### Actual Behaviour

Even after giving the "null" value, the module is still creating a **"network_security_group"** at virtual machine/NIC level.

**Comment by lonegunmanb** [comment]
Hi @anoopkrrish07 thanks for reporting this issue to us.

>By setting the "network_security_group" to "null", that should not create a new network security group.

Unfortunately that's not how `network_security_group` work. According to the [`description`](https://github.com/Azure/terraform-azurerm-compute/blob/5.2.2/variables.tf#L261):

>The network security group we'd like to bind with virtual machine. Set this variable will disable the creation of `azurerm_network_security_group` and `azurerm_network_security_rule` resources.

The reason why we wrapped nsg id with this object is, if you'd like to create your nsg in your root module, then pass the created nsg's id into this module, if we use nsg id directly in `count` expression, that would cause an error, Terraform don't allow using any expression that is `known after apply` in `count` for `for_each` expression. So we wrapped this id with an object, whether this object is `null` or not is known during the plan phase.

I'm sorry for the confusion, I hope this answer could solve your issue.
